### PR TITLE
Fix drawing artifacts due to rounding errors.

### PR DESCRIPTION
**Pull request by carlo-bramini**
This patch fixes issue #130.

I have seen that there are already some for fixing those graphic glitches, but in my opinion they are not exactly correct.
Increasing the value of `NSVG__FIXSHIFT` just hides the real cause of the problem.
In my opinion, there is a mathematical error into the conversion from the floating point value to fixed point.
Here, the code is using `floorf()` for converting the floating point value to integer, but I think that this is wrong because it does not take care of the remaining valid digits of that number.
Rounding to the nearest high or low integer value discards the value of those valid digits, not included into the fractional part of the fixed point number.
As result, this error is accumulated on each step and, at the end, the limit of the integral does not converge to the expected value.
That's why increasing `NSVG__FIXSHIFT` apparently fixes the issue, but in my opinion a better solution is to change how the integer value is converted and keeping `NSVG__FIXSHIFT` as it is, at least for now.

In other words, `floorf()` simply needs to be replaced with `roundf()`.
This patch has fixed the test case included into the issue #130 and probably other drawing misalignments.

### CODE REVIEW

**Comment by oehhar** [comment]
IMHO the patch looks great. Perhaps, @memononen has a minuite to verify.

To my experience, froundf is not available in all compilers. So ffloorf(x+0.5) should be used.
But this is memononen to decide.

Thank you !
Harald

**Comment by carlo-bramini** [comment]
Hello, thank you very much for your reply.
However:
> To my experience, froundf is not available in all compilers. So ffloorf(x+0.5) should be used. But this is memononen to decide.

this is not exactly correct, because `floor()` cannot be used as direct replacement of `round()`.
If the number is positive, you can use `floor(x+0.5)`.
But if the number is negative, you must do `ceil(x+0.5)`.
Or, if you prefer, `floor(x+0.5)` if it's positive and `floor(x-0.5)` if it's negative.
In my opinion, using `round()` or its float version `roundf()` is the right choice for fixing the issue, actually.

**Comment by wcout** [comment]
@carlo-bramini Not sure if this fixes things really or just with this image. I did a very short test with your fix with other angles of the line and the misplacement comes and goes depending on the angle...

**Comment by oehhar** [comment]
Carlo,
thanks for your comment and correction, that is great. To my knowledge, ANSI C89 did not contain round and floor should be used. The [Tksvg] project specifies to be compiled with MS-VC 6 which does not have round(). Due to that, I made this quite unrelated comment.
Take care,
Harald

**Comment by carlo-bramini** [comment]
The patch has been modified by introducing a function `nsvg__roundf()`, for supporting compilation under C89 and MSVC6.

**Comment by oehhar** [comment]
Hi Carlo,
I can confirm, that it compiles well with MS-VC++ 6. I imported your changes into tksvg and compiled.
The warnings (not your fault) are as follows:

```
C:\test\git\tksvg\win\..\generic\nanosvg.h(1325) : warning C4244: '=' : conversion from 'double ' to 'float ', possible loss of data
C:\test\git\tksvg\win\..\generic\nanosvg.h(1349) : warning C4244: '=' : conversion from 'float ' to 'unsigned int ', possible loss of data
C:\test\git\tksvg\win\..\generic\nanosvg.h(1350) : warning C4244: '=' : conversion from 'float ' to 'unsigned int ', possible loss of data
C:\test\git\tksvg\win\..\generic\nanosvg.h(1351) : warning C4244: '=' : conversion from 'float ' to 'unsigned int ', possible loss of data
```

When I replace
```
static float nsvg__roundf(float x) { return (x > 0) ? floorf(x + 0.5) : ceilf(x + 0.5); }
```
by
```
static float nsvg__roundf(float x) { return roundf(x); }
```
it compiles well, to. So, wether "roundf" is present in MS-VC 6++ or the sourrounding TCL/Tk framework has already a work-around for it.

My comment was not ment, that this is required. memononen is the only decider here. He often imports bug-fixes, but it may take time (months).

Perhaps, I may import your patch intk tksvg and tk itself. I normally prefer, that memononen first puts it to trunk. Nevertheless, we already have other pull requersts included, which are not committed to trunk.

It would be great to address the point by @wcount above.

Thank you for all your work !
Harald

**Comment by oehhar** [comment]
Another question. Above your change, there is a comment:

`/* round dx down to avoid going too far */`

Is this comment still correct? Is it still "round dx *down*?

Thank you,
Harald

**Comment by oehhar** [comment]
It is now included in Tk [Tk commit b75baea5cb1d7b5c](https://core.tcl-lang.org/tk/info/b75baea5cb1d7b5c) and in tksvg [tksvg https://github.com/oehhar/tksvg/tree/2690da71f70c48e0e20f8d1c50f86c6c2090cd3b](https://github.com/oehhar/tksvg/tree/2690da71f70c48e0e20f8d1c50f86c6c2090cd3b).

Thank you for all,
Harald

**Comment by carlo-bramini** [comment]
@oehhar I have been also able to make nanosvg working on MSVC6, but it requires some fixes.
I have seen that you did some also in your code, like here:
https://github.com/oehhar/tksvg/blob/2690da71f70c48e0e20f8d1c50f86c6c2090cd3b/generic/nanosvg.h#L1155

**Comment by oehhar** [comment]
Carlo,
MS-VC6 is not a requirement for this project. I highly regret, that I have mentioned it.
Nevertheless, great, that you have made it compile.
If you have any hints for tksvg and MS-VC6, I am willing to integrate them.

I would also revert your special solution to avoid roundf.

Sorry,
Harald

**Comment by memononen** [comment]
Thanks for the fix and persistence to follow it through, and thanks for the commenters too.

File: src/nanosvgrast.h [hunk 1]
```diff
@@ -331,6 +331,7 @@ static float nsvg__normalize(float *x, float* y)
 }
 
 static float nsvg__absf(float x) { return x < 0 ? -x : x; }
+static float nsvg__roundf(float x) { return (x > 0) ? floorf(x + 0.5) : ceilf(x + 0.5); }
```
Comment (sezero) [inline src/nanosvgrast.h:334]: Sorry for late review, but noticed this: I think the third part should be `ceilf(x - 0.5)` and not `ceilf(x + 0.5)`

Am I wrong?

Comment (memononen) [reply src/nanosvgrast.h:334]: My quick godbolt agrees with you, and also >= 0 so that positive zero stays positive.

Comment (memononen) [reply src/nanosvgrast.h:334]: Fixed in https://github.com/memononen/nanosvg/commit/cb0ae54e6b147ccdf85401ef3ef20f2c761252c0

Comment (sezero) [reply src/nanosvgrast.h:334]: Thanks
